include/envoy/stats/stats_macros.h

```diff
@@ -69,4 +102,49 @@ static inline std::string statPrefixJoin(absl::string_view prefix, absl::string_
 
 #define NULL_POOL_GAUGE(POOL) (POOL).nullGauge(NULL_STAT_DECL_IGNORE_MODE_
+
+// Used for declaring StatNames in a structure.
+#define GENERATE_STAT_NAME_STRUCT(NAME, ...) Envoy::Stats::StatName NAME##_;
+#define GENERATE_STAT_NAME_INIT(NAME, ...) , NAME##_(pool_.add(#NAME))
+
+// Macros for declaring stat-structures using StatNames, for those that must be
+// instantiated during operation, and where speed and scale matters. These
+// macros are not for direct use; they are only for use from
+// MAKE_STAT_NAMES_STRUCT. and MAKE_STAT_STRUCT.
+#define MAKE_STATS_STRUCT_COUNTER_HELPER_(NAME)                                                    \
+  , NAME##_(Envoy::Stats::Utility::counterFromStatNames(scope, {prefix, stat_names.NAME##_}))
+#define MAKE_STATS_STRUCT_GAUGE_HELPER_(NAME, MODE)                                                \
+  , NAME##_(Envoy::Stats::Utility::gaugeFromStatNames(scope, {prefix, stat_names.NAME##_},         \
+                                                      Envoy::Stats::Gauge::ImportMode::MODE))
+#define MAKE_STATS_STRUCT_STATNAME_HELPER_(name)
+#define GENERATE_STATNAME_STRUCT(name)
+
+/**
+ * Generates a struct with StatNames for a subsystem, based on the stats macro
+ * with COUNTER, GAUGE, HISTOGRAM, TEXT_READOUT, and STAT_NAME calls.
+ */
+#define MAKE_STAT_NAMES_STRUCT(StatNamesStruct, ALL_STATS)                                         \
+  struct StatNamesStruct {                                                                         \
+    explicit StatNamesStruct(Envoy::Stats::SymbolTable& symbol_table)                              \
+        : pool_(symbol_table) ALL_STATS(GENERATE_STAT_NAME_INIT, GENERATE_STAT_NAME_INIT,          \
+                                        GENERATE_STAT_NAME_INIT) {}                                \
+    Envoy::Stats::StatNamePool pool_;                                                              \
+    ALL_STATS(GENERATE_STAT_NAME_STRUCT, GENERATE_STAT_NAME_STRUCT, GENERATE_STAT_NAME_STRUCT)     \
+  }
+
+/**
+ * Instantiates a structure of stats based on a new scope and optional prefix, using
+ * a predefined structure of stat names.
+ */
+#define MAKE_STATS_STRUCT(StatsStruct, StatNamesStruct, ALL_STATS)                                 \
+  struct StatsStruct {                                                                             \
+    StatsStruct(const StatNamesStruct& stat_names, Envoy::Stats::Scope& scope,                     \
+                Envoy::Stats::StatName prefix = Envoy::Stats::StatName())                          \
+        : scope_(scope)                                                                            \
+              ALL_STATS(MAKE_STATS_STRUCT_COUNTER_HELPER_, MAKE_STATS_STRUCT_GAUGE_HELPER_,        \
+                        MAKE_STATS_STRUCT_STATNAME_HELPER_) {}                                     \
+    Envoy::Stats::Scope& scope_;                                                                   \
```

Comment:
Does holding the scope reference here actually do anything?

Reply:
No. this was done for syntactic purposes only as I needed the macro to have a leading comma. I should add comment about that unfortunate hack.

Reply:
I wound up replacing the scope here (which we do not need) with a ref to the StatNames struct, which turns out to be useful (though not essential).